### ISSUE TITLE
CPU-side ColorJitter is a dataloader bottleneck — could the GPU path be restored?

### ISSUE THREAD

**Issue by xiaojunjie** [NVIDIA/cosmos-framework#174]
DROID action SFT currently runs ColorJitter inside the dataloader workers, which is
computationally expensive.

1. In principle: ColorJitter is a random augmentation to begin with, so moving it to
   the GPU should only introduce CPU/CUDA kernel-level numerical differences, without
   changing the training semantics — is that right?

2. In practice: ColorJitter and the tail after it can be moved out of the dataloader
   worker and run in the main process on the GPU. The pipeline stage and the ordering stay exactly
   the same; only the execution device changes. The worker also passes along its RNG
   state, which keeps the final difference limited to CPU/CUDA kernel rounding.
   The cost is that worker-to-main-process traffic doubles, since the tensor returned
   by the worker goes from [T, C, 3H/2, W] to [3T, C, H, W].

Does this approach look viable?

**Comment by lfengad** [comment]
Yes to (1), with one caveat, and (2) works but costs more than you estimated.

**(1)** Correct — it's a random augmentation, so the device swap is a re-parameterization, not a semantic change. But don't ship the worker's RNG *state*: ColorJitter draws from the global torch RNG, so restoring it in the main process clobbers the stream the model uses (flow-matching noise, `cfg_dropout_rate`, dropout). Ship the sampled parameters instead — 4 factors plus the op-order permutation. Also keep them **per sample**; jittering the collated batch once would cut color diversity by `batch_size`.

**(2)** Viable, but there are two costs:

- The worker doesn't return the composite — `_convert_video` casts to uint8 and `VideoResize` resizes/pads after it. Your pre-jitter tensor is float32, so it's **~7×**, not 2×, unless the uint8 cast also moves ahead of the jitter. And since the jitter precedes composition, compose + convert + resize/pad all have to move with it.
- More importantly, this trades a CPU bottleneck for **GPU critical path + IPC bandwidth**. The GPU is the scarce resource in this recipe, and moving the jitter there adds an RGB→HSV→RGB pass to every step, while IPC traffic goes up 2–7×. That may just relocate the stall from the worker into the step rather than remove it.

Cheaper first step, same reasoning: **reorder instead of relocating**. Run the jitter after composition — L/R are already downscaled 2×, so that's exactly half the pixels — and after the uint8 cast, for another 4× off the memory traffic. Few-line diff, no IPC change. Worth profiling `hue=0.08` first though; it's the one op with no integer fast path and likely most of the cost.

**Comment by harshitwandhare** [comment]
Profiled the reorder on CPU (single op, no dataloader overhead), transforms.v2 0.20.1, 16 frames at 180x320 per view:

| pipeline | dtype | avg / call |
| --- | --- | --- |
| current (jitter before compose, on wrist+left+right concatenated) | float32 | 148.5 ms |
| reordered (jitter after compose+downscale) | float32 | 80.2 ms |
| reordered | uint8 | 107.7 ms |

Reorder alone: 1.85x, close to the 2x the pixel count predicts (composite is 1.5*H*W per frame against 3*H*W for the three raw views, since left/right are already downscaled by the time jitter runs).

The uint8 step doesn't hold up here though: casting to uint8 first makes the reordered path 1.35x *slower*, not 4x faster. `ColorJitter` in this torchvision version round-trips uint8 through float internally for the hue/saturation ops (RGB<->HSV), so the memory-traffic argument doesn't pay off, at least not on CPU, where the conversion overhead eats the bandwidth win. Worth checking whether that changes on GPU, but the dataloader workers here are CPU-bound, which is the thing actually being profiled.

So: reorder is a real, verified ~1.85x on the jitter op itself, uint8 first is not, at least as `ColorJitter` is implemented today. Planning to send a PR for the reorder only (`_compose_multi_view` in `droid_lerobot_dataset.py`), skip the uint8 step, and note in the docstring why compose happens before jitter now. Will keep the RNG-state fix from your first comment: sample and pass a per-item `(brightness, contrast, saturation, hue, order)` tuple rather than restoring worker RNG state.

**Comment by harshitwandhare** [comment]
Sent the reorder as #219.

Measured on `_compose_multi_view` itself rather than an isolated ColorJitter call, so the number includes the crop, resize, interpolate and cat that did not change. 16 frames at 180x320 per view, float32, `torch.set_num_threads(4)`, 20 calls per round after a warmup, with the two variants alternating round by round so load drift hits both sides equally:

```
round 1: before=208.2  after=115.0
round 2: before=223.7  after=128.9
round 3: before=209.6  after=135.7
round 4: before=222.7  after=129.6
round 5: before=216.3  after=119.1

median before = 216.3 ms, median after = 128.9 ms, ratio = 1.68x
max(after) = 135.7 ms, min(before) = 208.2 ms
```

**About 1.7x on the whole method**, sets non-overlapping.

Correcting myself: this comment first said 1.9x, from three runs of one side followed by three of the other. That attributes any load drift to whichever side ran second. Alternating the rounds gives 1.7x and that is the number to use. At default thread count the same method reads 1.84x.

One thing worth flagging on your RNG point. It applies to the relocation path, not to this one: the reorder keeps the jitter inside the worker, so no parameters cross a process boundary and nothing touches the main process RNG stream. Your "keep them per sample" caveat is also already satisfied here, since `_compose_multi_view` runs per `__getitem__` and draws once per sample, not once per collated batch. So there is nothing to carry over from that half of the advice into this PR, which is why you will not see the parameter-passing change I mentioned I would keep.

Crop and resize had to stay per view, since cropping the composite would cut across the tile boundaries, so only the colour op moved.

Left #174 open rather than marking it closed by the PR, since the GPU relocation question you were originally asked about is still unanswered and this only takes the cheaper first step.

**Comment by harshitwandhare** [comment]
The reorder half of this landed as #219, merged today.

It went in gated rather than as the default: @pengcuo added a `jitter_after_compose` flag, off by default, so the shipped pipeline is byte-identical to before (same-seed sha256 matched 15/15 across three shapes and five seeds) until an internal A/B decides whether to flip it. With the flag on, their measurement is 169.3 ms to 100.3 ms, about 1.69x, which matches the 1.68x I measured on different hardware.

They also produced the number I could not: same-seed PSNR at or above 49.4 dB against 12.8 dB for the augmentation itself, so the reorder shifts the output by roughly 0.5% of the augmentation's own magnitude.

This issue's actual question is still open. Moving the jitter to the GPU is the larger win and nobody has measured it. On the costs @lfengad set out above, the GPU path prices at roughly 7x rather than the 2x the original framing assumed, once the critical path and IPC are counted, but that is arithmetic on his figures and not a measurement. I have no CUDA machine, so I cannot settle it here.

Also still standing from earlier in the thread: the uint8-first half of the proposal measured 1.35x slower in this torchvision version, so it is not worth carrying into any GPU version without re-checking on the target stack.
